src/Reflection/Apply.agda

```diff
@@ -0,0 +1,200 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Function application of reflected terms
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Data.Nat
+
+module Reflection.Apply (limit : ℕ) where
+
+open import Level renaming (suc to lsuc; _⊔_ to lmax)
+open import Data.List
+open import Data.Sum.Base
+open import Data.Product
+open import Function.Base
+open import Data.Nat.Base
+open import Relation.Nullary
+open import Data.Bool.Base
+open import Reflection hiding (map-Args ; returnTC ; _≟_ ; _>>=_) renaming (return to returnTC)
+open import Reflection.Argument
+open import Reflection.Argument.Visibility using () renaming (_≟_ to _≟ᵥ_)
+open import Reflection.Pattern hiding (_≟_)
+open import Reflection.Term
+open import Data.Maybe.Base hiding (_>>=_)
+
+Result : Set 0ℓ → Set 0ℓ
+Result A = List ErrorPart ⊎ A
+
+pattern ok v = inj₂ v
+pattern err e = inj₁ e
+
+module _ where
+  import Data.Sum.Categorical.Left (List ErrorPart) 0ℓ as S
+  open import Category.Monad
+  open RawMonad S.monad
+
+  failed-app : Term → Arg Term → Result Term
+  failed-app t (arg _ a) = err (strErr "attempted to apply" ∷ termErr a ∷ strErr "to" ∷ termErr t ∷ [])
+
+  data Fuel : Set 0ℓ where
+    fuel : ℕ → Fuel
+
+  shift-index : ℕ → ℕ → Term → Term
+  shift-index-args : ℕ → ℕ → Args Term → Args Term
+  shift-index-clauses : ℕ → ℕ → List Clause → List Clause
```

Comment:
A couple of comments about what these functions do would be awesome!

Reply:
Agreed.